std/assembly/array.ts

```diff
@@ -476,6 +476,64 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 1 : child.length;
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    let byteLength = size << usize(alignof<valueof<T>>());
+    let dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value
+    let result: valueof<T>[] = changetype<valueof<T>[]>(__alloc(offsetof<valueof<T>[]>(), idof<valueof<T>[]>()));
+    store<i32>(changetype<usize>(result), size, offsetof<valueof<T>[]>("length_"));
```

Comment:
Can a manual store be avoided by typing `result` as `Array<valueof<T>>` maybe?

Reply:
No. This is because `length_` is private.

Reply:
Still wondering because the `.length_` property is read above, but on `this`. Is this a TypeScript limitation that another function of the same class cannot access is other than on `this`? Or is this maybe a type mismatch between TS Array and AS Array that can be solved with a `changetype<Array<valueof<T>>>` above (with `Array` being the AS array)?

Reply:
I cannot access the private _length of the target. I can only access private length_ of `this`. I would say that is by design, yes.

Reply:
The following does not trigger a TS error

```ts
class Foo {
  private bar = 1;
  func(other: Foo): void {
    other.bar = 2;
  }
}
```

which makes me assume it should work here as well. Issue is that `something[]` is resolved differently than `Array<something>`. Have you tried my suggestion?

Reply:
These are all my assumptions :). Ive been out with the family all day. I don't mean to be stubborn.

Reply:
Yeah. I had no idea private property access was valid between instances. I had assumed the error in my dev environment I received was related to this assumption. It should be all fixed now. Sorry about the wait.